src/vm-vcpu-ref/src/aarch64/regs/mod.rs

```
@@ -0,0 +1,244 @@
+
+fn set_reg_data<Reg, RegChunk>(
+    fd: &DeviceFd,
+    reg: &Reg,
+    group: u32,
+    data: &GicRegState<RegChunk>,
+    mpidr: u64,
+    mpidr_mask: u64,
+) -> Result<()>
+where
+    Reg: MmioReg,
+    RegChunk: Clone,
+{
+    for (offset, val) in reg.iter::<RegChunk>().zip(&data.chunks) {
+        let mut tmp = (*val).clone();
+        fd.set_device_attr(&kvm_device_attr(group, offset, &mut tmp, mpidr, mpidr_mask))?;
+    }
+
+    Ok(())
+}
+
+fn get_regs_data<'a, Reg, RegChunk>(
+    fd: &DeviceFd,
+    regs: impl Iterator<Item = &'a Reg>,
+    group: u32,
+    mpidr: u64,
+    mpidr_mask: u64,
+) -> Result<Vec<GicRegState<RegChunk>>>
+where
+    Reg: MmioReg + 'a,
+    RegChunk: Default,
+{
+    let mut data = Vec::new();
+    for reg in regs {
+        data.push(get_reg_data(fd, reg, group, mpidr, mpidr_mask)?);
+    }
+
+    Ok(data)
+}
+
+fn get_reg_data<Reg, RegChunk>(
+    fd: &DeviceFd,
+    reg: &Reg,
+    group: u32,
+    mpidr: u64,
+    mpidr_mask: u64,
+) -> Result<GicRegState<RegChunk>>
+where
+    Reg: MmioReg,
+    RegChunk: Default,
+{
+    let mut data = Vec::with_capacity(reg.iter::<RegChunk>().count());
+    for offset in reg.iter::<RegChunk>() {
+        let mut val = RegChunk::default();
+        fd.get_device_attr(&mut kvm_device_attr(
+            group, offset, &mut val, mpidr, mpidr_mask,
+        ))?;
+        data.push(val);
+    }
+
+    Ok(GicRegState { chunks: data })
+}
+
+fn kvm_device_attr<RegChunk>(
+    group: u32,
+    offset: u64,
+    val: &mut RegChunk,
+    mpidr: u64,
+    mpidr_mask: u64,
+) -> kvm_device_attr {
+    kvm_device_attr {
+        group,
+        attr: (mpidr & mpidr_mask) | offset,
+        addr: val as *mut RegChunk as u64,
+        flags: 0,
+    }
+}
+
+trait MmioReg {
+    fn range(&self) -> Range<u64>;
+
+    fn iter<T>(&self) -> StepBy<Range<u64>>
+    where
+        Self: Sized,
+    {
+        self.range().step_by(std::mem::size_of::<T>())
+    }
+}
+
+#[derive(PartialEq)]
+struct SimpleReg {
```

Can you add some comments about what is a simple reg vs other types of regs?

---

Given the current code structure, I couldn't find a better way than to have `Option<Gic>` and initialize it later. Yes, this indeed can panic, but that would be a clear mistake, and I'd prefer to fail fast instead of handling it as just another error condition.

---

That makes sense right now, we might change that in the future if we want to export finer grained control. I would still add a comment about the panic & why it is not panic-ing at the moment as it's not immediately clear. Btw, I think this response was for a different comment?

---

Yes you're right :)